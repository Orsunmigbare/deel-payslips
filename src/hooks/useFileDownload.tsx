import { useState } from "react";
import { Capacitor } from "@capacitor/core"
import { Filesystem, Directory } from '@capacitor/filesystem';
import { FileOpener } from "@capacitor-community/file-opener"
import { base64FromPath, downloadFile } from "src/utilities/fileDownload";


export const useFileDownload = () => {
    const [loading, setLoading] = useState(false);

    const download = async (url: string, filename: string, onComplete?: () => void, onError?: () => void) => {
        if (Capacitor.isNativePlatform()) {
            try {
                setLoading(true);
                const base64 = await base64FromPath(url);
                const saveFile = await Filesystem.writeFile({
                    path: filename,
                    data: base64,
                    directory: Directory.Data
                })

                await FileOpener.open({
                    filePath: saveFile.uri
                })
                setLoading(false);
                onComplete?.();
            } catch (err) {
                setLoading(false);
                onError?.()
            }

        } else {
            downloadFile(url, filename);
        }
    }

    return {
        download,
        loading,
    }
}