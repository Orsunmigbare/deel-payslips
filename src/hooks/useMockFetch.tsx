import { useState, useEffect } from "react";
import { fetchPayslip, fetchPayslips } from "utilities/mockAPI"


type MockFetchState = {
    loading: boolean;
    data?: any
}

export const useMockFetch = (type: string, id?: string) => {
    const [state, setState] = useState<MockFetchState>({
        loading: false,
    })

    const fetchData = async (type: string) => {
        setState({ ...state, loading: true })
        let response;
        if (type === "payslip" && id) {
            response = await fetchPayslip(id);
        } else {
            response = await fetchPayslips()
        }
        setState({ loading: false, data: response.data })
    }

    useEffect(() => {
        fetchData(type)
    }, [])

    return {
        loading: state.loading,
        data: state.data
    }
}