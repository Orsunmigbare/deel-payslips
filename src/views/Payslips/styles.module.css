.hDesc {
    font-size: 12px;
    padding-left: 18px;
}