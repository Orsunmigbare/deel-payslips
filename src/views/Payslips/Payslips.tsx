import { Card } from "components/Card/Card";
import {
    IonPage,
    IonContent,
    IonHeader,
    IonTitle,
    IonToolbar,
    IonList,
} from '@ionic/react';
import { Payslip as IPayslip } from 'types/api';
import { useMockFetch } from "hooks/useMockFetch";
import { Conditional } from "components/Conditional/Conditional";
import styles from "./styles.module.css"


type FetchResponse = {
    loading: boolean;
    data: IPayslip[]
}

function Payslips() {
    const { loading, data: slips }: FetchResponse = useMockFetch("payslips");
    return (
        <IonPage>
            <IonHeader translucent={true}>
                <IonToolbar>
                    <IonTitle>Payslips</IonTitle>
                </IonToolbar>
            </IonHeader>
            <IonContent className="ion-padding" fullscreen={true}>
                <IonHeader collapse="condense">
                    <IonToolbar>
                        <IonTitle size="large">Payslips</IonTitle>
                    </IonToolbar>
                </IonHeader>
                <h2 className={styles.hDesc}>View available payslips in your history</h2>
                <Conditional data={slips} loading={loading}>
                    <IonList>
                        {slips?.map((slip) =>
                            <Card key={slip.id} {...slip} />
                        )}
                    </IonList>
                </Conditional>

            </IonContent>
        </IonPage>
    )
}

export default Payslips
