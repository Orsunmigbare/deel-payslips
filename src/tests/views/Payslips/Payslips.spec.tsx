import "@testing-library/jest-dom"
import { render } from '@testing-library/react';
import { useMockFetch } from "src/hooks/useMockFetch";
import Payslips from "src/views/Payslips/Payslips";
import { sampleSlips } from "./data";

jest.mock("../../../hooks/useMockFetch", () => (
    {
        useMockFetch: jest.fn()
    }
));

afterEach(() => {
    jest.resetAllMocks();
});

describe("Payslips", () => {
    it("renders correctly", () => {
        (useMockFetch as jest.Mock).mockReturnValue({ loading: false, data: sampleSlips });
        const screen = render(<Payslips />);
        sampleSlips.forEach((slip, i) => {
            const durationText = `Payslip for ${new Date(slip.fromDate).toDateString()} to ${new Date(slip.toDate).toDateString()}`
            expect(screen.getByText(durationText)).toBeInTheDocument();
            expect(screen.getAllByAltText('payslip')[i]).toHaveAttribute("src", slip.file);
            expect(document.querySelectorAll("ion-item")[i]).toHaveAttribute("href", `/payslip/${slip.id}`)
        })
        expect(screen).toMatchSnapshot()
    })
})