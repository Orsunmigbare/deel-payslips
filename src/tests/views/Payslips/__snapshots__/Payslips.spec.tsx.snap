// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Payslips renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ion-page"
      >
        <ion-header
          translucent="true"
        >
          <!---->
          <ion-toolbar>
            <ion-title>
              Payslips
            </ion-title>
          </ion-toolbar>
        </ion-header>
        <ion-content
          class="ion-padding"
          fullscreen="true"
        >
          <ion-header
            collapse="condense"
          >
            <!---->
            <ion-toolbar>
              <ion-title
                size="large"
              >
                Payslips
              </ion-title>
            </ion-toolbar>
          </ion-header>
          <h2
            class="hDesc"
          >
            View available payslips in your history
          </h2>
          <ion-list>
            <ion-item
              class="cardCont"
              detail="false"
              href="/payslip/61f4a809a6c0b5d0e1840c11"
              lines="none"
            >
              <div
                class="card"
              >
                <div
                  class="info"
                >
                  <h3
                    class="infoTitle"
                  >
                    Payslip for 
                    Tue Feb 01 2022
                     to 
                    Tue Mar 01 2022
                  </h3>
                  <div
                    class="infoDesc"
                  >
                    Click to for details
                  </div>
                </div>
                <div
                  class="preview"
                >
                  <ion-img
                    alt="payslip"
                    class="previewImg"
                    src="https://www.personalfinancelab.com/wp-content/uploads/h5p/content/13/images/image-5e502af138e62.PNG"
                  />
                </div>
              </div>
            </ion-item>
            <ion-item
              class="cardCont"
              detail="false"
              href="/payslip/61f4a80aa6c0b5d0e1840c12"
              lines="none"
            >
              <div
                class="card"
              >
                <div
                  class="info"
                >
                  <h3
                    class="infoTitle"
                  >
                    Payslip for 
                    Tue Mar 01 2022
                     to 
                    Fri Apr 01 2022
                  </h3>
                  <div
                    class="infoDesc"
                  >
                    Click to for details
                  </div>
                </div>
                <div
                  class="preview"
                >
                  <ion-img
                    alt="payslip"
                    class="previewImg"
                    src="https://www.personalfinancelab.com/wp-content/uploads/h5p/content/13/images/image-5e502af138e62.PNG"
                  />
                </div>
              </div>
            </ion-item>
          </ion-list>
        </ion-content>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="ion-page"
    >
      <ion-header
        translucent="true"
      >
        <!---->
        <ion-toolbar>
          <ion-title>
            Payslips
          </ion-title>
        </ion-toolbar>
      </ion-header>
      <ion-content
        class="ion-padding"
        fullscreen="true"
      >
        <ion-header
          collapse="condense"
        >
          <!---->
          <ion-toolbar>
            <ion-title
              size="large"
            >
              Payslips
            </ion-title>
          </ion-toolbar>
        </ion-header>
        <h2
          class="hDesc"
        >
          View available payslips in your history
        </h2>
        <ion-list>
          <ion-item
            class="cardCont"
            detail="false"
            href="/payslip/61f4a809a6c0b5d0e1840c11"
            lines="none"
          >
            <div
              class="card"
            >
              <div
                class="info"
              >
                <h3
                  class="infoTitle"
                >
                  Payslip for 
                  Tue Feb 01 2022
                   to 
                  Tue Mar 01 2022
                </h3>
                <div
                  class="infoDesc"
                >
                  Click to for details
                </div>
              </div>
              <div
                class="preview"
              >
                <ion-img
                  alt="payslip"
                  class="previewImg"
                  src="https://www.personalfinancelab.com/wp-content/uploads/h5p/content/13/images/image-5e502af138e62.PNG"
                />
              </div>
            </div>
          </ion-item>
          <ion-item
            class="cardCont"
            detail="false"
            href="/payslip/61f4a80aa6c0b5d0e1840c12"
            lines="none"
          >
            <div
              class="card"
            >
              <div
                class="info"
              >
                <h3
                  class="infoTitle"
                >
                  Payslip for 
                  Tue Mar 01 2022
                   to 
                  Fri Apr 01 2022
                </h3>
                <div
                  class="infoDesc"
                >
                  Click to for details
                </div>
              </div>
              <div
                class="preview"
              >
                <ion-img
                  alt="payslip"
                  class="previewImg"
                  src="https://www.personalfinancelab.com/wp-content/uploads/h5p/content/13/images/image-5e502af138e62.PNG"
                />
              </div>
            </div>
          </ion-item>
        </ion-list>
      </ion-content>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
